pkg/controllers/node/node_controller.go

```diff
@@ -193,3 +308,155 @@ func kick(c chan<- interface{}) {
+	expectedIPs := []string{}
+	if node.Spec.BGP != nil {
+		// BGP IPv4 and IPv6 addresses are CIDRs.
+		if node.Spec.BGP.IPv4Address != "" {
+			ip, _, _ := net.ParseCIDROrIP(node.Spec.BGP.IPv4Address)
+			expectedIPs = append(expectedIPs, ip.String())
+		}
+		if node.Spec.BGP.IPv6Address != "" {
+			ip, _, _ := net.ParseCIDROrIP(node.Spec.BGP.IPv6Address)
+			expectedIPs = append(expectedIPs, ip.String())
+		}
+		if node.Spec.BGP.IPv4IPIPTunnelAddr != "" {
+			expectedIPs = append(expectedIPs, node.Spec.BGP.IPv4IPIPTunnelAddr)
+		}
+	}
+	if node.Spec.IPv4VXLANTunnelAddr != "" {
+		expectedIPs = append(expectedIPs, node.Spec.IPv4VXLANTunnelAddr)
+	}
+	return expectedIPs
+}
+
+// generateHostendpointFromNode returns the expected auto hostendpoint to be
+// created from the given node.
+func (c *NodeController) generateHostendpointFromNode(node *api.Node) *api.HostEndpoint {
+	hepLabels := make(map[string]string, len(node.Labels)+1)
+	for k, v := range node.Labels {
+		hepLabels[k] = v
+	}
+	hepLabels[hepCreatedLabelKey] = hepCreatedLabelValue
+
+	return &api.HostEndpoint{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   c.generateHostendpointName(node.Name),
+			Labels: hepLabels,
+		},
+		Spec: api.HostEndpointSpec{
+			Node:          node.Name,
+			InterfaceName: "*",
+			ExpectedIPs:   c.getHostendpointExpectedIPs(node),
+		},
+	}
+}
+
+// hostendpointNeedsUpdate returns true if the current automatic hostendpoint
+// needs to be updated.
+func (c *NodeController) hostendpointNeedsUpdate(current *api.HostEndpoint, expected *api.HostEndpoint) bool {
+	if !reflect.DeepEqual(current.Labels, expected.Labels) {
+		return true
+	}
+	if !reflect.DeepEqual(current.Spec.ExpectedIPs, expected.Spec.ExpectedIPs) {
+		return true
+	}
+	return current.Spec.InterfaceName != expected.Spec.InterfaceName
+}
+
+// updateHostendpoint updates the current hostendpoint so that it matches the
+// expected hostendpoint.
+func (c *NodeController) updateHostendpoint(current *api.HostEndpoint, expected *api.HostEndpoint) error {
+	expected.ResourceVersion = current.ResourceVersion
+	expected.ObjectMeta.CreationTimestamp = current.ObjectMeta.CreationTimestamp
+	expected.ObjectMeta.UID = current.ObjectMeta.UID
+	_, err := c.calicoClient.HostEndpoints().Update(context.Background(), expected, options.SetOptions{})
+	return err
+}
+
+// syncHostendpoint syncs the auto hostendpoint for the given node.
+func (c *NodeController) syncHostendpoint(node *api.Node) {
+	// On failure, we retry a certain number of times.
+	for n := 0; n < 5; n++ {
+		// Try getting the host endpoint.
+		hepName := c.generateHostendpointName(node.Name)
+		currentHep, err := c.calicoClient.HostEndpoints().Get(c.ctx, hepName, options.GetOptions{})
+		expectedHep := c.generateHostendpointFromNode(node)
+
+		// If the hostendpoint doesn't exist, create it.
+		if currentHep == nil {
+			switch err.(type) {
+			case errors.ErrorResourceDoesNotExist:
```

Comment:
Hm, this is actually a little bit confusing (having the error handling from above only within this block). 

We should probably handle the error at the same point that we do the `Get` call.

Reply:
refactored in aa0a5ca